### PR TITLE
ARROW-10024: [C++][Parquet] Create nested reading benchmarks

### PR DESCRIPTION
**Pull request by pitrou**
These benchmarks only measure one-level nesting (struct, list) for now.

### CODE REVIEW

**Comment by pitrou** [comment]
Here are the results here:
```
BM_ReadStructColumn/0     1706505 ns      1704955 ns          409 bytes_per_second=6.87335G/s items_per_second=615.017M/s
BM_ReadStructColumn/1     7827665 ns      7821031 ns           90 bytes_per_second=1.49836G/s items_per_second=134.071M/s
BM_ReadStructColumn/50   14142868 ns     14135647 ns           49 bytes_per_second=848.918M/s items_per_second=74.1796M/s
BM_ReadStructColumn/99    3846944 ns      3844119 ns          182 bytes_per_second=3.04849G/s items_per_second=272.774M/s

BM_ReadListColumn/0     7949690 ns      7946976 ns           87 bytes_per_second=1006.67M/s items_per_second=131.947M/s
BM_ReadListColumn/1     9302755 ns      9299473 ns           76 bytes_per_second=860.264M/s items_per_second=112.756M/s
BM_ReadListColumn/50   19390741 ns     19386146 ns           36 bytes_per_second=412.666M/s items_per_second=54.0889M/s
BM_ReadListColumn/99    7162830 ns      7159609 ns           97 bytes_per_second=1117.38M/s items_per_second=146.457M/s
```

**Comment by pitrou** [comment]
@ursabot archery benchmark run --suite-filter=parquet-arrow-reader-writer-benchmark

**Comment by pitrou** [comment]
@ursabot benchmark run --suite-filter=parquet-arrow-reader-writer-benchmark

**Comment by kszucs** [comment]
@ursabot benchmark run --help

**Comment by pitrou** [comment]
@ursabot archery benchmark --suite-filter=parquet-arrow-reader-writer-benchmark

**Comment by pitrou** [comment]
@ursabot benchmark --suite-filter=parquet-arrow-reader-writer-benchmark

**Comment by wesm** [comment]
Looks like there is something wrong with the Ursabot result reporting, but you can see the results here:

https://ci.ursalabs.org/#/builders/73/builds/103/steps/3/logs/result

**Comment by pitrou** [comment]
Yeah, that's not what I wanted anyway. I want to run the new benchmarks, not diff existing benchmarks. Nevermind...

File: cpp/src/arrow/testing/random.cc [hunk 1]
```diff
@@ -250,12 +267,20 @@ std::shared_ptr<Array> RandomArrayGenerator::StringWithRepeats(int64_t size,
...
   using GenOpt = GenerateOptions<int32_t, std::uniform_int_distribution<int32_t>>;
-  GenOpt options(seed(), first_offset, last_offset, /*null_probability=*/0);
+  GenOpt options(seed(), first_offset, last_offset, null_probability);
 
   BufferVector buffers{2};
 
+  int64_t null_count = 0;
+  buffers[0] = *AllocateEmptyBitmap(size);
+  options.GenerateBitmap(buffers[0]->mutable_data(), size, &null_count);
+  // Make sure the first and last entry are non-null
```
Comment (emkornfield) [inline cpp/src/arrow/testing/random.cc:280]: why is this important?

File: cpp/src/arrow/testing/random.h [hunk 2]
```diff
@@ -52,7 +60,7 @@ class ARROW_TESTING_EXPORT RandomArrayGenerator {
   std::shared_ptr<Array> Boolean(int64_t size, double true_probability,
                                  double null_probability = 0);
 
-  /// \brief Generates a random UInt8Array
+  /// \brief Generate a random UInt8Array
```
Comment (emkornfield) [inline cpp/src/arrow/testing/random.h:63]: nit: why the change?  If we want to standardize on one vs the other we should probably update the style guide?

File: cpp/src/parquet/arrow/reader_writer_benchmark.cc [hunk 3]
```diff
@@ -280,6 +298,81 @@ BENCHMARK_TEMPLATE2(BM_ReadColumn, true, BooleanType)
...
+static void BM_ReadStructColumn(::benchmark::State& state) {
+  constexpr int64_t kNumValues = BENCHMARK_SIZE / 10;
+  const double null_probability = static_cast<double>(state.range(0)) / 100.0;
+  const bool nullable = (null_probability != 0.0);
+
+  ARROW_CHECK_GE(null_probability, 0.0);
+
+  ::arrow::random::RandomArrayGenerator rng(42);
+
+  auto values1 = rng.Int32(kNumValues, -5, 5, null_probability);
```
Comment (emkornfield) [inline cpp/src/parquet/arrow/reader_writer_benchmark.cc:314]: could you comment, the -5, and 5 to indicate what these parameters represent?

File: cpp/src/parquet/arrow/reader_writer_benchmark.cc [hunk 4]
```diff
@@ -280,6 +298,81 @@ BENCHMARK_TEMPLATE2(BM_ReadColumn, true, BooleanType)
...
+  constexpr int64_t kNumValues = BENCHMARK_SIZE / 10;
+  const double null_probability = static_cast<double>(state.range(0)) / 100.0;
+  const bool nullable = (null_probability != 0.0);
+
+  ARROW_CHECK_GE(null_probability, 0.0);
+
+  ::arrow::random::RandomArrayGenerator rng(42);
+
+  auto values = rng.Int64(kNumValues, -5, 5, null_probability);
+  auto offsets = rng.Offsets(kNumValues / 10, 0, values->length(), null_probability);
```
Comment (emkornfield) [inline cpp/src/parquet/arrow/reader_writer_benchmark.cc:353]: approximately what size lists does this generate?

Comment (pitrou) [reply cpp/src/parquet/arrow/reader_writer_benchmark.cc:353]: `kNumValues / 10`

File: cpp/src/arrow/testing/random.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/testing/random.cc:280]: Hmm... you're right, it's probably not :-S

File: cpp/src/parquet/arrow/reader_writer_benchmark.cc (hunk 4 above)
Comment (emkornfield) [reply cpp/src/parquet/arrow/reader_writer_benchmark.cc:353]: so lists get larger as overall array goes up?

Comment (pitrou) [reply cpp/src/parquet/arrow/reader_writer_benchmark.cc:353]: Ah, you meant the size of individual lists? Well, since the list array will be `kNumValues / 10` large, individual lists should have 10 items average.

Comment (emkornfield) [reply cpp/src/parquet/arrow/reader_writer_benchmark.cc:353]: ah, OK, I'm not sure if this is representative of real workloads but it is better then nothing.

File: cpp/src/parquet/arrow/reader_writer_benchmark.cc (hunk 3 above)
Comment (pitrou) [reply cpp/src/parquet/arrow/reader_writer_benchmark.cc:314]: Will do.

File: cpp/src/parquet/arrow/reader_writer_benchmark.cc (hunk 4 above)
Comment (pitrou) [reply cpp/src/parquet/arrow/reader_writer_benchmark.cc:353]: It's not really meant to be representative of anything real, just not outlandish :-)

File: cpp/src/arrow/testing/random.h (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/testing/random.h:63]: Hmm, I can add this somewhere in the docs. Most of our docstrings use infinite, AFAIK.

File: cpp/src/arrow/testing/random.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/testing/random.cc:280]: Well, at least the last offset should be non-null (according to `ListArray::FromArrays`). The first one needn't, apparently.

Comment (pitrou) [reply cpp/src/arrow/testing/random.cc:280]: I'd rather keep this after all. It doesn't hurt and at least half of it is necessary.

**Comment by pitrou** [comment]
I think I addressed your review comments @emkornfield .

**Comment by emkornfield** [comment]
Yes + 1 thanks.

**Comment by emkornfield** [comment]
merging

**Comment by pitrou** [comment]
Thanks for the review :-)
